File: algorithms/[111]二叉树的最小深度/solution.py
# topics = ["树", "广度优先搜索"]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from __future__ import annotations
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, x: int = 0, left: Optional[TreeNode] = None, right: Optional[TreeNode] = None):
        self.val = x
        self.left = left
        self.right = right


class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        res = 0
        queue = deque()
        queue.append(root)

        while queue:
            res += 1
            # 一次处理一层
            for _ in range(len(queue)):
                node = queue.popleft()
                if not node.left and not node.right:
                    return res
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

        return res

File: algorithms/[111]二叉树的最小深度/solution.ts
// topics = ["树", "深度优先搜索"]

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function minDepth(root: TreeNode | null): number {
  const helper = (node: TreeNode, k: number): number => {
    if (!node.left && !node.right) {
      return k;
    }
    if (!node.left) {
      return helper(node.right!, k + 1);
    }
    if (!node.right) {
      return helper(node.left!, k + 1);
    }
    return Math.min(helper(node.left, k + 1)!, helper(node.right, k + 1));
  };

  if (!root) {
    return 0;
  }
  return helper(root, 1);
}
